Step 1: Import Required Libraries

In [50]:
# Data Handling
import pandas as pd
import numpy as np

# Train-Test Split
from sklearn.model_selection import train_test_split

# Pipeline
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# Preprocessing
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

# Machine Learning Model
from sklearn.ensemble import RandomForestClassifier

# Evaluation
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Save Model
import joblib

In [51]:
df = pd.read_csv("/content/drive/MyDrive/CSV/WA_Fn-UseC_-Telco-Customer-Churn.csv")

Step 2: Import the Dataset

In [52]:
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/CSV/WA_Fn-UseC_-Telco-Customer-Churn.csv")

## Display First Five Rows of the Dataset

In [53]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


Observation :
The dataset contains customer demographic information, subscribed services, billing details, and the target variable Churn.
Each row represents one telecom customer.
The target column Churn indicates whether a customer has left the service (Yes) or remained with the company (No).
The first five rows confirm that the dataset has been loaded successfully and is ready for further preprocessing.

Check the Shape of the Dataset

In [54]:
df.shape

(7043, 21)

Interpretation: 
7043 → Total number of customer records (rows).
21 → Total number of features/columns in the dataset.

Observation:

The dataset contains 7,043 customer records and 21 columns. Each row represents a unique customer, while the columns contain information such as customer demographics, subscribed services, billing details, and the target variable Churn. This confirms that the dataset has sufficient records for building and evaluating a machine learning model.

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


The dataset contains 7,043 records and 21 columns.
There are 18 categorical (object) columns, 2 integer (int64) columns, and 1 floating-point (float64) column.
All columns have 7,043 non-null values, indicating that there are no missing values at this stage. However, the TotalCharges column has the object data type because it contains blank spaces (" "), which need to be converted into numeric values during data preprocessing.
The dataset occupies approximately 1.1 MB of memory.

In [57]:
(df["TotalCharges"] == " ").sum()

np.int64(11)

Observation :

The TotalCharges column contains 11 blank space values (" ").
These blank spaces are not treated as missing values (NaN) by Pandas.
Before converting the column to a numeric data type, these blank spaces need to be replaced with NaN.
This preprocessing step is necessary to avoid errors during model training and numerical operations.

Replace Blank Spaces with Missing Values (NaN)

In [58]:
df["TotalCharges"] = df["TotalCharges"].replace(" ", np.nan)

There is no visible output because the replacement is performed directly on the DataFrame.

Observation
All blank spaces (" ") in the TotalCharges column have been replaced with NaN.
The missing values can now be detected using functions such as isnull().
This step prepares the data for numerical conversion and missing value imputation in the preprocessing pipeline.

# Convert TotalCharges to Numeric Data Type

In [59]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"])

In [60]:
print(df["TotalCharges"].dtype)
print(df["TotalCharges"].isnull().sum())

float64
11


Observation :

The TotalCharges column has been successfully converted from object to float64.
There are 11 missing values (NaN) in the column after conversion.
These missing values will be handled later using the Median Imputer in the preprocessing pipeline.
The column is now ready for machine learning preprocessing and model training.

# Remove the customerID Column

In [62]:
# Remove customerID column
df.drop("customerID", axis=1, inplace=True)

There is no visible output because the column is removed directly from the original DataFrame using inplace=True.

Observation :

The customerID column has been successfully removed from the dataset.
The remaining columns contain only the features that are relevant for customer churn prediction.
Removing unnecessary identifier columns helps improve the quality and efficiency of the machine learning model.

# Verify the Remaining Columns

In [63]:
df.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

Observation :

The customerID column has been successfully removed from the dataset.
The dataset now contains 20 columns, including 19 input features and 1 target variable (Churn).
All remaining columns are relevant for building the Customer Churn Prediction model.

# Train-Test Split

In [64]:
# Features
X = df.drop("Churn", axis=1)

# Target
y = df["Churn"]

In [65]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [66]:
print(X_train.shape)
print(X_test.shape)

(5634, 19)
(1409, 19)


The dataset contains 19 input features after removing the target column (Churn).
The training dataset consists of 5,634 records, which will be used to train the model.
The testing dataset consists of 1,409 records, which will be used to evaluate the model.
Using stratified sampling ensures that both datasets have a similar distribution of churn and non-churn customers, helping to produce reliable and unbiased model evaluation.

# Feature Classification

In [67]:
# Numerical features
numeric_features = [
    "SeniorCitizen",
    "tenure",
    "MonthlyCharges",
    "TotalCharges"
]

# Categorical features
categorical_features = [
    "gender",
    "Partner",
    "Dependents",
    "PhoneService",
    "MultipleLines",
    "InternetService",
    "OnlineSecurity",
    "OnlineBackup",
    "DeviceProtection",
    "TechSupport",
    "StreamingTV",
    "StreamingMovies",
    "Contract",
    "PaperlessBilling",
    "PaymentMethod"
]

Observation:


A total of 4 numerical features were identified:
SeniorCitizen
tenure
MonthlyCharges
TotalCharges
A total of 15 categorical features were identified:
gender
Partner
Dependents
PhoneService
MultipleLines
InternetService
OnlineSecurity
OnlineBackup
DeviceProtection
TechSupport
StreamingTV
StreamingMovies
Contract
PaperlessBilling
PaymentMethod
The total number of input features is 19, matching the dataset after removing the target variable (Churn).

# Numerical Data Preprocessing Pipeline

In [68]:
numeric_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

Missing values in numerical features will be replaced using the median.
After imputation, numerical features will be standardized using StandardScaler.
Combining preprocessing steps into a pipeline ensures that the same transformations are consistently applied to both the training and testing datasets.
This approach also reduces the risk of data leakage and makes the preprocessing workflow more organized and reproducible.

# Categorical Data Preprocessing Pipeline

In [69]:
categorical_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

Missing values in categorical features are handled using the most frequent category (mode).
Categorical variables are transformed into numerical format using One-Hot Encoding.
handle_unknown="ignore" improves model robustness by preventing errors when new or unseen categories appear in the test dataset.
This preprocessing ensures that machine learning models can interpret categorical data effectively.

Step 6.4: Column Transformer

In [70]:
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_pipeline, numeric_features),
        ("cat", categorical_pipeline, categorical_features)
    ]
)

Explanation

Ye automatically:

Numerical columns → Numerical Pipeline
Categorical columns → Categorical Pipeline

Step 7: Final Random Forest Pipeline

In [72]:
model_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model", RandomForestClassifier(
        n_estimators=100,
        random_state=42
    ))
])

Step 8: Train the Model

In [73]:
model_pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['SeniorCitizen', 'tenure',
                                                   'MonthlyCharges',
                                                   'TotalCharges']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('encoder',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['gender', 'Partner',
                                                   'Dependents', 'PhoneService',
                                                   'MultipleLines',
                                                   'InternetService',
                                                   'OnlineSecurity',
                                                   'OnlineBackup',
                                                   'DeviceProtection',
                                                   'TechSupport', 'StreamingTV',
                                                   'StreamingMovies',
                                                   'Contract',
                                                   'PaperlessBilling',
                                                   'PaymentMethod'])])),
                ('model', RandomForestClassifier(random_state=42))])

The pipeline successfully combines preprocessing and model training in a single workflow.
Both numerical and categorical features are automatically transformed before training.
The model used here is RandomForestClassifier, which is powerful for classification tasks like churn prediction.
Training the pipeline ensures that the same preprocessing steps will automatically be applied during prediction, preventing data leakage.
The structure improves code reusability, simplicity, and reliability.

Step 9: Make Predictions

In [75]:
# Make predictions
y_pred = model_pipeline.predict(X_test)

In [76]:
X_test

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges
437,Male,0,Yes,Yes,72,Yes,Yes,Fiber optic,Yes,Yes,Yes,Yes,Yes,Yes,Two year,Yes,Credit card (automatic),114.05,8468.20
2280,Female,1,No,No,8,Yes,Yes,Fiber optic,No,No,No,Yes,Yes,Yes,Month-to-month,Yes,Credit card (automatic),100.15,908.55
2235,Female,0,Yes,Yes,41,Yes,Yes,DSL,Yes,Yes,Yes,No,Yes,No,One year,Yes,Credit card (automatic),78.35,3211.20
4460,Male,0,Yes,No,18,Yes,No,Fiber optic,No,No,Yes,Yes,No,No,Month-to-month,No,Electronic check,78.20,1468.75
3761,Female,0,Yes,No,72,Yes,Yes,DSL,Yes,Yes,Yes,No,Yes,Yes,Two year,Yes,Credit card (automatic),82.65,5919.35
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5143,Female,0,Yes,Yes,49,Yes,No,DSL,Yes,Yes,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,87.20,4345.00
4439,Male,0,Yes,Yes,28,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,Yes,Credit card (automatic),20.30,487.95
3857,Male,0,No,No,5,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Bank transfer (automatic),20.65,93.55
4758,Female,0,No,No,56,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Bank transfer (automatic),19.70,1051.90


The model successfully predicts outcomes for all 1,409 test samples.
X_test contains 19 original features, but internally they are transformed into a higher-dimensional format due to One-Hot Encoding.
The pipeline ensures that preprocessing is applied consistently before prediction.
This step produces the final output (y_pred), which will later be compared with y_test to evaluate model performance.

# Viewing Predictions

In [77]:
print(y_pred[:10])

['No' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes']


The model predicts a mix of "Yes" and "No" outcomes.
Most predictions in this sample are "No", indicating that many customers are expected to stay.
A few "Yes" predictions show that the model is able to identify potential churn cases.
This step helps quickly verify that the model is producing meaningful outputs before formal evaluation.

Step 10: Calculate Accuracy

In [78]:
# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)

Accuracy: 0.7778566359119943


The model achieved an accuracy of ~77.79%.
This means the model correctly predicts customer churn status for approximately 78 out of 100 customers.
The result indicates that the model has a good baseline performance, but there is still room for improvement.
Since churn prediction is often an imbalanced and business-critical problem, accuracy alone may not be sufficient for full evaluation.

# Model Evaluation (Accuracy Score)

In [79]:
y_pred = model_pipeline.predict(X_test)

In [80]:
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.7778566359119943


The model achieved an accuracy of 0.7778 (~77.8%).
This indicates that the model correctly predicts churn status for nearly 78% of customers in the test dataset.
The pipeline (preprocessing + RandomForestClassifier) is performing consistently on unseen data.
However, accuracy alone is not enough for a full evaluation, especially in churn prediction problems where class imbalance may exist.

Step 11: Confusion Matrix

In [81]:
# Confusion Matrix
cm = confusion_matrix(y_test, y_pred)

print(cm)

[[918 117]
 [196 178]]


The model correctly identifies a large number of non-churning customers (918 TN).
It also correctly detects 178 churn cases (TP).
However, there are:
117 false alarms (FP)
196 missed churn cases (FN) → this is important in business context
The model performs better at predicting "No Churn" than "Churn".

Step 12: Classification Report

In [82]:
# Classification Report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          No       0.82      0.89      0.85      1035
         Yes       0.60      0.48      0.53       374

    accuracy                           0.78      1409
   macro avg       0.71      0.68      0.69      1409
weighted avg       0.77      0.78      0.77      1409



The model performs better for "No churn" class (0.85 F1-score) compared to "Yes churn" (0.53 F1-score).
High recall for No (0.89) → model is good at identifying customers who will stay.
Low recall for Yes (0.48) → model is missing many churn customers.
Overall accuracy remains ~78%, but class-wise performance is uneven.

Step 13: Save the Model

In [83]:
# Save the trained model
joblib.dump(model_pipeline, "customer_churn_random_forest.pkl")

['customer_churn_random_forest.pkl']

The entire ML pipeline is successfully saved as a .pkl file.
This ensures portability and reusability of the trained model.
The saved model can now be loaded anytime using:

# Loading Saved Model & Making Predictions

In [84]:
loaded_model = joblib.load("customer_churn_random_forest.pkl")

In [85]:
prediction = loaded_model.predict(X_test)

The saved model is successfully reloaded without retraining.
Predictions are generated directly from the stored pipeline.
This confirms that the model is:
Properly saved
Reusable
Deployment-ready